src/suricata.c

```diff
@@ -1112,6 +1142,37 @@ static int ParseCommandLineAfpacket(SCInstance *suri, const char *in_arg)
 #endif
 }
 
+static int ParseCommandLineDpdk(SCInstance *suri, const char *in_arg)
+{
+#ifdef HAVE_DPDK
+    if (suri->run_mode == RUNMODE_UNKNOWN) {
+        suri->run_mode = RUNMODE_DPDK;
+        if (in_arg) {
+            LiveRegisterDeviceName(in_arg);
+            memset(suri->pcap_dev, 0, sizeof(suri->pcap_dev));
+            strlcpy(suri->pcap_dev, in_arg, sizeof(suri->pcap_dev));
+        }
+    } else if (suri->run_mode == RUNMODE_DPDK) {
+        if (in_arg) {
+            LiveRegisterDeviceName(in_arg);
+        } else {
+            SCLogInfo("Multiple dpdk options without interface on each is useless");
+        }
+    } else {
+        SCLogError(SC_ERR_MULTIPLE_RUN_MODE, "more than one run mode "
+                                             "has been specified");
+        PrintUsage(suri->progname);
+        return TM_ECODE_FAILED;
+    }
+    return TM_ECODE_OK;
+#else
+    SCLogError(SC_ERR_NO_DPDK, "DPDK not enabled. On Linux "
```

Comment:
Now this is new information **DPDK bypass uses Suricata as a secondary process**. question

1. is Windows and openBSD no supported documented in YAML?
2. is Windows and openBSD no supported documented in PDF suricata?
3. I do not see your check for check if secondary process after `rte_eal_init` (rework)
4. where is the sample application for DPDK running as primary?
5. how are you taking care of ASLR? are you recommending to disable?
6. where is `base address` option set to prevent ASLR?

Did @victorjulien  agree on running DPDK-Suricata as a secondary process? because @victorjulien  had suggested worker mode and primary dpdk application .

Reply:
From the point of view of the suricata integration, what are the consequences of supporting primary and secondary process modes of operation?

Reply:
I would like to emphasize that in this PR (and the following versions of this PR), Suricata is considered only as a primary DPDK process and works in workers mode. 

What DPDK mode is used to run Suricata is rather conceptual. It can be implemented both ways. 

Consequences of supporting Suricata to run as a primary DPDK process:
- Suricata must allocate hugepages - only the primary process can allocate hugepages. 
- Suricata is responsible to configure the NICs according to the configuration stored in the suricata.yaml file. This requires the implementation of all bits of code to support device configuration (I think the configuration might be done by the secondary process but it is rather common for the primary process).
- if other DPDK applications running as a secondary process want to connect to Suricata, Suricata is responsible to:
  - create DPDK rings for applications to interconnect Suricata with applications. Via these rings, Suricata could pass packets to secondary apps and vice versa.
  - facilitate a way of communication to allocate hugepages for secondary processes. Since secondary applications can not allocate hugepages alone, they would need to send a request to Suricata to do it for them. Suricata can then reply with the result of this operation. For this use case, possibly a [DPDK IPC](https://doc.dpdk.org/guides/prog_guide/multi_proc_support.html) or a ring structure could be used.
- secondary applications can be stopped/changed during the Suricata runtime
  

Consequences of supporting Suricata to run as a secondary DPDK process:
- configuration is abstracted from Suricata, the NICs are already preconfigured by the primary process
- memory allocation is abstracted from Suricata, Suricata only attaches to RX/TX DPDK rings. This way, Suricata can be connected with the primary process. Apart from the DPDK rings, Suricata does not require to allocate more hugepages (therefore, there is no need to create a communication channel between the primary and the secondary processes)
- if Suricata runs as a secondary process, all NICs must be handled by the primary application (the primary application must create DPDK rings for all NICs). 
- Suricata can be connected in a pipeline model with other applications, e.g. the primary process can create rings in a way that one application pushes packets to rx rings, another application detects DDoS attacks and then Suricata analyzes the traffic. Packets are moved between these applications in DPDK rings. 


These are the consequences, advantages, and disadvantages of differences between primary and secondary DPDK processes off the top of my head. There can be many more.
But I believe we could discuss this somewhere else (e.g. Redmine) and at a later point in time. Current PR focuses on Suricata running in workers mode and as a primary process, exactly as Victor suggested. I would not like to steer the main point of the pull request. 
The work I mentioned in the comment above can be a topic for further discussion. But until then, DPDK Suricata does not contain a built-in DPDK bypass method and as of now, we can ignore the fact that Suricata can be a secondary process later on.

Reply:
Thanks for the explanation. I'm initially interested in Suricata as the primary process, but in the longer term I think it makes sense to make both modes possible. So it would be my preference if the initial implementation would be done in such a way that extending it later wouldn't be too hard.